### DISCUSSION TITLE
Utility to emit assets during build

### DISCUSSION THREAD

**Discussion by natemoo-re** [withastro/roadmap#492] [Proposal]
### Body

# Summary

There are many reasons runtime code might want to emit an asset during our build process. The official `@astrojs/image` integration or a possible `@astrojs/og` integration are two recent examples. 

Currently, similar patterns are possible, assuming you have a very deep knowledge of Astro, Vite, and our entire build process. We should generalize the best practices and provide a baked-in utility for this.

# Background & Motivation

There have been many community integrations which write files directly to `public/` as an attempt to generate files during the build, but this is not recommended. See discussion in https://github.com/withastro/astro/pull/5854.

As I was building an Open Graph image integration, I referenced `@astrojs/image` a lot, but the logic is very complex and has to account for `static` and `server` outputs.

A generic utility could abstract all of this complex logic away, making it fairly easy to reason about assets generated in the build. Integrations and components would get automatic dev + build and static + server behavior for free.

# Goals

A **concise, bulleted-list** outlining the intended goals of this RFC.

- Provide a simple interface to allow integrations/components to emit assets during the `astro build` process.
- Abstract away the complexities of managing asset generation between `static` and `server` modes, `dev` and `build` environments, and `prerender` mode.
- Unify logic between official integrations that need to do things like this.

# Example

Hand-wavey example, but something like this could map well to our existing build process and be access from any component.

```js
import { emitAsset } from 'astro:bundle';

const generatedAssetPath = await emitAsset('@namespace/asset.[hash].json', JSON.stringify(data))
```

**Comment by Princesseuh** [comment]
+1 to an approved way of doing this. As you said, the image integration has to do a complex dance to account for all the situations.

Would caching of the generated assets be in scope? (As in, "don't build assets if already built")

**Reply by natemoo-re** [reply]
Yep, totally in scope for this! Here's a spike on an `emitAssetWithCache` utility

https://github.com/natemoo-re/astro-bundle-demo/blob/main/src/components/Cached.astro#L11-L14

**Comment by natemoo-re** [comment]
Interesting development... my exploration here worked great for `static` mode and local dev, but got super interesting when I prototyped `server` mode. Take a look at my [branch](https://github.com/natemoo-re/astro-bundle-demo/pull/2) / [demo](https://deploy-preview-2--fanciful-kulfi-2dc331.netlify.app/).

On the server, we obviously can't write to the file system or rely on persistent memory, so we need to persist our data externally. I decided to prototype with [`keyv`](https://www.npmjs.com/package/keyv) and realized that it actually works very nicely. Users would own their data, database, and hosting, but could simply provide a URL:

```
redis://user:pass@localhost:6379
mongodb://user:pass@localhost:27017/dbname
sqlite://path/to/database.sqlite
postgresql://user:pass@localhost:5432/dbname
mysql://user:pass@localhost:3306/dbname
etcd://localhost:2379
```

On the Astro side, we can use `injectRoute` to generate an endpoint that reads from this database and returns an immutable `Response`. This would remove a lot of the pain points with our current server approach, where assets always need to be built on Request.

**Reply by matthewp** [reply]
What does this look like from the integration developer side? How do I tell it that I want to use one of these namespaced URLs?

**Comment by tony-sull** [comment]
Would love to see us standardize this!  Caching was a late addition to the current `@astrojs/images` integration and a feature that was begging for a second use case to hammer out a proper standardized API

Instead of building this into Astro itself, could this actually be part of our integrations API?  I know Netlify and Vercel bubble up access for adding files to their shared build cache, I wonder if we could tie into that to automatically cache built assets after the first request 🤔

**Comment by florian-lefebvre** [comment]
I believe this is now covered by the `astro:build:done` hook: https://docs.astro.build/en/reference/integrations-reference/#dir-option-1

**Reply by delucis** [reply]
I think I’ll reopen this one as I don’t agree — a component can’t use that hook really and the aim here was to provide a simple API that even components could use. (Having access to `dir` is also not really the same as having access to a nice `emitAsset()` utility.)

The value here would presumably be the encapsulation of logic so it can also work at runtime for SSR sites, which is the complex part.

**Comment by delucis** [comment]
@natemoo-re’s work shared here was super helpful! I just published an integration that reuses much of that logic: https://delucis.github.io/astro-emit-asset/

It’s static only for now (won’t work in an on-demand route), but it would probably be possible to use platform-specific KV stores to support on-demand usage in theory (via unstorage similar to how sessions work). I’m not too interested in the on-demand side myself, so won’t investigate that immediately though.
